### PR TITLE
Group by document and not editor

### LINKED ISSUE #316: Cannot issue `move` command from one editor to the other if they are both on the same document

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
fixes #316

### CODE REVIEW

File: src/util/targetUtils.ts [hunk 1]
```diff
@@ -29,10 +29,14 @@ export async function runForEachEditor<T, U>(
...
 ): Promise<U[]> {
+  // Actually group by document and not editor. If the same document is open in multiple editors we want to perform all actions in one editor or an concurrency error will occur.
+  const getDocument = (target: T) => getEditor(target).document;
+  const editorMap = groupBy(targets, getDocument);
   return await Promise.all(
-    Array.from(groupBy(targets, getEditor), async ([editor, editorTargets]) =>
-      func(editor, editorTargets)
-    )
+    Array.from(editorMap.values(), async (editorTargets) => {
+      const editor = getEditor(editorTargets[0]);
```
Comment (pokey) [inline src/util/targetUtils.ts:37]:
```suggestion
      # Just pick any editor with the given document open; doesn't matter which
      const editor = getEditor(editorTargets[0]);
```

**Review by pokey — COMMENTED** [review]
Looks good, but please add a test

**Comment by pokey** [comment]
To my knowledge, we have no tests that operate on two editors with the same document, which is what this PR is supposed to fix.  Would be good to have a test for that

**Comment by AndreasArvidsson** [comment]
To my knowledge we can't do tests like that. The code changes are used by almost every single action so we have regression testing. The issue with multiple editors for the same document is tested manually and I have no idea how to write a test for that.

**Comment by pokey** [comment]
Yeah you can't use test case recorder but it shouldn't be too bad.  Have a look at https://github.com/pokey/cursorless-vscode/blob/14749f4f42b30fe8c7f7c9906ce79138982b0b2e/src/test/suite/backwardCompatibility.test.ts#L21-L36

**Comment by AndreasArvidsson** [comment]
The only way to open the same document in multiple editors is to actually open a file on disk. A bit problematic on running unit tests isn't it?

**Comment by pokey** [comment]
> The only way to open the same document in multiple editors is to actually open a file on disk. A bit problematic on running unit tests isn't it?

Can just make a temp file

**Comment by AndreasArvidsson** [comment]
Of course true. I might have a look at that when I have some time over.

**Comment by pokey** [comment]
I'm not sure I want us to get in the habit of merging PRs without tests. I made an exception for #401 because it required #160, which is out of scope for that PR. Not sure I see why a test is out of scope for this PR

**Comment by AndreasArvidsson** [comment]
@pokey Test added. I took special care to make sure that the hats are now in different editors of the same document.

File: src/test/suite/groupByDocument.test.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  const hat1 = hatMap
+    .getEntries()
+    .find(([, token]) => token.editor === editor1 && token.text === "hello");
+  const hat2 = hatMap
+    .getEntries()
+    .find(([, token]) => token.editor === editor2 && token.text === "world");
+
+  const [color1, char1] = hat1![0].split(".");
+  const [color2, char2] = hat2![0].split(".");
```
Comment (pokey) [inline src/test/suite/groupByDocument.test.ts:44]: **nit**: Please use [`HatTokenMap.splitKey`](https://github.com/pokey/cursorless-vscode/blob/6cb9dea7a6baacd9a7245b67ada7593270941850/src/core/HatTokenMap.ts#L57)

**Review by pokey — COMMENTED** [review]
Awesome test thank you!  Left a tiny nit but otherwise looks great

**Review by pokey — APPROVED** [review]
looks good!

**Pull request merged**